File: backend/app/schemas/learning.py
from pydantic import BaseModel
from uuid import UUID
from datetime import datetime
from typing import Optional, List
from app.schemas.skill import SkillResponse
from app.schemas.task import TaskResponse

class RoleSkillRequirement(BaseModel):
    skill_id: UUID
    skill_name: str
    required_level: float
    importance: str

class RoleResponse(BaseModel):
    id: UUID
    name: str
    category: Optional[str] = None
    description: Optional[str] = None
    requirements: List[RoleSkillRequirement] = []

    class Config:
        from_attributes = True

class SkillGapItem(BaseModel):
    skill_id: UUID
    skill_name: str
    category: Optional[str] = None
    current_level: float
    required_level: float
    gap: float
    importance: str
    severity: str  # CRITICAL, IMPORTANT, MODERATE, MINOR
    recommended_action: str

class SkillGapReport(BaseModel):
    role_id: UUID
    role_name: str
    overall_readiness: float
    total_gaps: int
    critical_gaps: int
    gaps: List[SkillGapItem]

class LearningResourceResponse(BaseModel):
    id: UUID
    title: str
    provider: Optional[str] = None
    url: Optional[str] = None
    description: Optional[str] = None
    difficulty: str

    class Config:
        from_attributes = True

class LearningPathItemResponse(BaseModel):
    id: UUID
    learning_path_id: UUID
    skill_id: UUID
    skill_name: Optional[str] = None
    title: str
    sequence_number: int
    status: str
    progress: float
    resource: Optional[LearningResourceResponse] = None

    class Config:
        from_attributes = True

class LearningPathResponse(BaseModel):
    id: UUID
    user_id: UUID
    goal_id: Optional[UUID] = None
    role_id: Optional[UUID] = None
    title: str
    status: str
    created_at: datetime
    items: List[LearningPathItemResponse] = []

    class Config:
        from_attributes = True

class GenerateRoadmapRequest(BaseModel):
    role_id: Optional[UUID] = None
    goal_id: Optional[UUID] = None
    title: Optional[str] = None

class ConvertLearningToTaskResponse(BaseModel):
    task: TaskResponse
    message: str

File: frontend/src/services/api/learning.ts
import { apiClient } from './client';
import { Task } from '@/types';

export interface RoleSkillRequirement {
  skill_id: string;
  skill_name: string;
  required_level: number;
  importance: string;
}

export interface Role {
  id: string;
  name: string;
  category?: string;
  description?: string;
  requirements: RoleSkillRequirement[];
}

export interface SkillGapItem {
  skill_id: string;
  skill_name: string;
  category?: string;
  current_level: number;
  required_level: number;
  gap: number;
  importance: string;
  severity: 'CRITICAL' | 'IMPORTANT' | 'MODERATE' | 'MINOR';
  recommended_action: string;
}

export interface SkillGapReport {
  role_id: string;
  role_name: string;
  overall_readiness: number;
  total_gaps: number;
  critical_gaps: number;
  gaps: SkillGapItem[];
}

export interface LearningResource {
  id: string;
  title: string;
  provider?: string;
  url?: string;
  description?: string;
  difficulty: string;
}

export interface LearningPathItem {
  id: string;
  learning_path_id: string;
  skill_id: string;
  skill_name?: string;
  title: string;
  sequence_number: number;
  status: 'PENDING' | 'IN_PROGRESS' | 'COMPLETED';
  progress: number;
  resource?: LearningResource;
}

export interface LearningPath {
  id: string;
  user_id: string;
  goal_id?: string;
  role_id?: string;
  title: string;
  status: 'ACTIVE' | 'COMPLETED' | 'ARCHIVED';
  created_at: string;
  items: LearningPathItem[];
}

export const learningApi = {
  getRoles: () => apiClient<Role[]>('/roles'),

  getSkillGaps: (roleId?: string) => {
    const qs = roleId ? `?role_id=${roleId}` : '';
    return apiClient<SkillGapReport>(`/skills/gaps${qs}`);
  },

  getResources: () => apiClient<LearningResource[]>('/learning/resources'),

  getPaths: () => apiClient<LearningPath[]>('/learning/paths'),

  generateRoadmap: (data: { role_id?: string; goal_id?: string; title?: string }) =>
    apiClient<LearningPath>('/learning/paths/generate', {
      method: 'POST',
      body: JSON.stringify(data),
    }),

  convertToTask: (itemId: string) =>
    apiClient<{ task: Task; message: string }>(`/learning/items/${itemId}/convert-to-task`, {
      method: 'POST',
    }),

  completeItem: (itemId: string) =>
    apiClient<LearningPathItem>(`/learning/items/${itemId}/complete`, {
      method: 'POST',
    }),
};
